secretsmanager.tf

```diff
@@ -0,0 +1,273 @@
+    "${formatlist("%s:%s", data.aws_iam_role.saml_role.unique_id, var.secrets_saml_users)}",
+  ]
+
+  # list of role users and saml users for policies
+  role_and_saml_ids = [
+    "${aws_iam_role.app_role.unique_id}:*",
+    "${local.saml_user_ids}",
+  ]
+
+  sm_arn = "arn:aws:secretsmanager:${var.region}:${data.aws_caller_identity.current.account_id}:secret:${var.app}-${var.environment}-??????"
+}
+
+# create the KMS key for this secret
+resource "aws_kms_key" "sm_kms_key" {
+  description = "${var.app}-${var.environment}"
+  policy      = "${data.aws_iam_policy_document.kms_resource_policy_doc.json}"
+  tags        = "${merge(var.tags, map("Name", format("%s-%s", var.app, var.environment)))}"
+}
+
+# alias for the key
+resource "aws_kms_alias" "sm_kms_alias" {
+  name          = "alias/${var.app}-${var.environment}"
+  target_key_id = "${aws_kms_key.sm_kms_key.key_id}"
+}
+
+# the kms key policy
+data "aws_iam_policy_document" "kms_resource_policy_doc" {
+  statement = {
+    sid    = "DenyWriteToAllExceptSAMLUsers"
+    effect = "Deny"
+
+    principals = {
+      type        = "AWS"
+      identifiers = ["*"]
+    }
+
+    actions   = ["${local.kms_write_actions}"]
+    resources = ["*"]
+
+    condition = {
+      test     = "StringNotLike"
+      variable = "aws:userId"
+      values   = ["${local.saml_user_ids}"]
+    }
+  }
+
+  statement = {
+    sid    = "DenyReadToAllExceptRoleAndSAMLUsers"
+    effect = "Deny"
+
+    principals = {
+      type        = "AWS"
+      identifiers = ["*"]
+    }
+
+    actions   = ["${local.kms_read_actions}"]
+    resources = ["*"]
+
+    condition = {
+      test     = "StringNotLike"
+      variable = "aws:userId"
+      values   = ["${local.role_and_saml_ids}"]
+    }
+  }
+
+  statement = {
+    sid    = "AllowWriteToSAMLUsers"
+    effect = "Allow"
+
+    principals = {
+      type        = "AWS"
+      identifiers = ["*"]
+    }
+
+    actions   = ["${local.kms_write_actions}"]
+    resources = ["*"]
+
+    condition = {
+      test     = "StringLike"
+      variable = "aws:userId"
+      values   = ["${local.saml_user_ids}"]
+    }
+  }
+
+  statement = {
+    sid    = "AllowReadRoleAndSAMLUsers"
+    effect = "Allow"
+
+    principals = {
+      type        = "AWS"
+      identifiers = ["*"]
+    }
+
+    actions   = ["${local.kms_read_actions}"]
+    resources = ["*"]
+
+    condition = {
+      test     = "StringLike"
+      variable = "aws:userId"
+      values   = ["${local.role_and_saml_ids}"]
+    }
+  }
+}
+
+# create the secretsmanager secret
+resource "aws_secretsmanager_secret" "sm_secret" {
+  name       = "${var.app}-${var.environment}"
+  kms_key_id = "${aws_kms_key.sm_kms_key.key_id}"
+  tags       = "${var.tags}"
+  policy     = "${data.aws_iam_policy_document.sm_resource_policy_doc.json}"
+}
+
+# create the placeholder secret json
+resource "aws_secretsmanager_secret_version" "initial" {
+  secret_id     = "${aws_secretsmanager_secret.sm_secret.id}"
+  secret_string = "{}"
```

Comment:
Does this need a lifecycle/ignore_changes attribute?

Reply:
not since it's a version